hwy/contrib/unroller/unroller-inl.h

```diff
@@ -0,0 +1,496 @@
+    if (places < 0) mask = maskneg;
+
+    hn::BlendedStore(x, mask, d_out, to + idx);
+    return std::abs(places);
+  }
+
+  inline ptrdiff_t Reduce(hn::Vec<OT> const& HWY_RESTRICT x,
+                          OUT_T* HWY_RESTRICT to) {
+    return me()->ReduceImpl(x, to);
+  }
+
+  inline ptrdiff_t ReduceImpl(hn::Vec<OT> const& HWY_RESTRICT x, OUT_T* to) {
+    // default does nothing
+    return 0;
+  }
+
+  inline void Reduce(hn::Vec<OT> const& HWY_RESTRICT x0,
+                     hn::Vec<OT> const& HWY_RESTRICT x1,
+                     hn::Vec<OT> const& HWY_RESTRICT x2,
+                     hn::Vec<OT>& HWY_RESTRICT y) {
+    me()->ReduceImpl(x0, x1, x2, y);
+  }
+
+  inline void ReduceImpl(hn::Vec<OT> const& HWY_RESTRICT x0,
+                         hn::Vec<OT> const& HWY_RESTRICT x1,
+                         hn::Vec<OT> const& HWY_RESTRICT x2,
+                         hn::Vec<OT>& HWY_RESTRICT y) {
+    // default does nothing
+  }
+};
+
+template <class FUNC, typename IN_T, typename OUT_T>
+inline void Unroller(FUNC& f, IN_T* HWY_RESTRICT x, OUT_T* HWY_RESTRICT y,
+                     const ptrdiff_t n) {
+  const auto lane_sz = f.UnitLanes();
+
+  auto xx = f.X0Init();
+  auto yy = f.YInit();
+  ptrdiff_t i = 0;
+
+#if HWY_MEM_OPS_MIGHT_FAULT
+  if (n < lane_sz) {
+    // stack is maybe too small for this in RVV?
+    IN_T xtmp[lane_sz];
+    OUT_T ytmp[lane_sz];
+
+    memcpy(xtmp, x, n * sizeof(IN_T));
+    xx = f.MaskLoad(0, xtmp, n);
+    yy = f.Func(0, xx, yy);
+    i += f.MaskStore(0, ytmp, yy, n);
+    i += f.Reduce(yy, ytmp);
+    memcpy(y, ytmp, i * sizeof(OUT_T));
+    return;
+  }
+#endif
+
+  if (n > 4 * lane_sz) {
+    auto xx1 = f.X0Init();
+    auto yy1 = f.YInit();
+    auto xx2 = f.X0Init();
+    auto yy2 = f.YInit();
+    auto xx3 = f.X0Init();
+    auto yy3 = f.YInit();
+
+    while (i + 4 * lane_sz - 1 < n) {
+      xx = f.Load(i, x);
+      i += lane_sz;
+      xx1 = f.Load(i, x);
+      i += lane_sz;
+      xx2 = f.Load(i, x);
+      i += lane_sz;
+      xx3 = f.Load(i, x);
+      i -= 3 * lane_sz;
+
+      yy = f.Func(i, xx, yy);
+      yy1 = f.Func(i + lane_sz, xx1, yy1);
+      yy2 = f.Func(i + 2 * lane_sz, xx2, yy2);
+      yy3 = f.Func(i + 3 * lane_sz, xx3, yy3);
+
+      f.Store(i, y, yy);
+      i += lane_sz;
+      f.Store(i, y, yy1);
+      i += lane_sz;
+      f.Store(i, y, yy2);
+      i += lane_sz;
+      f.Store(i, y, yy3);
+      i += lane_sz;
+    }
+
+    f.Reduce(yy3, yy2, yy1, yy);
+  }
+
+  while (i + lane_sz - 1 < n) {
+    xx = f.Load(i, x);
+    yy = f.Func(i, xx, yy);
+    f.Store(i, y, yy);
+    i += lane_sz;
+  }
+
+  if (i != n) {
+    xx = f.MaskLoad(n - lane_sz, x, i - n);
+    yy = f.Func(n - lane_sz, xx, yy);
+    f.MaskStore(n - lane_sz, y, yy, i - n);
+  }
+
+  f.Reduce(yy, y);
+}
+
+template <class FUNC, typename IN0_T, typename IN1_T, typename OUT_T>
+inline void Unroller(FUNC& HWY_RESTRICT f, IN0_T* HWY_RESTRICT x0,
```

Comment:
Are we expecting FUNC to record any state? If not, it would be convenient for callers if they could pass FUNC(), which would require the f arg here to be const&.

Also, if the reducer always returns one value, can we have that returned from this function? That would allow `const T result = Unroller();` rather than `T result; Unroller(.., &result)`.

Reply:
> Are we expecting FUNC to record any state?

I don't think it will record state. I'm having difficulty thinking of a reason it would. It can contain state, but I don't think there's a reason to allow it to alter state. I was trying to do this a couple days ago, and I encountered a problem. I don't exactly remember what the problem was, but I had to make all the member functions `const`, and I vaguely remember something like the `D d` member variables of the class being passed into non-const highway functions was causing the builds to fail. Does that sound right?

> if the reducer always returns one value

It returns one value in the examples I made (I'll make more), but it returns `void` in non-reducer loops.